File: reactor-game/backend/handlers/bonuses.go
package handlers

import (
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"reactor-game/backend/models"
	"strconv"
	"time"

	"github.com/jmoiron/sqlx"
)

type BonusResponse struct {
	Balance       int            `json:"balance"`
	ActiveReactor models.Reactor `json:"active_reactor"`
	FarmStatus    string         `json:"farm_status"`
	Progress      int            `json:"progress"`
	TimeLeft      int            `json:"time_left"` //в секундах
}

func GetBonuses(db *sqlx.DB) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		//1. Находим пользователя
		userIDStr := r.URL.Query().Get("userID")
		userID, err := strconv.Atoi(userIDStr)
		if err != nil {
			http.Error(w, "Invalid id", http.StatusBadRequest)
			return
		}
		log.Print("Trying to get Bonuses for user with id=", userID)
		var user models.User
		err = db.Get(&user, "SELECT * FROM users WHERE id=$1", userID)
		if err != nil {
			http.Error(w, "user not found", http.StatusNotFound)
			log.Fatal(err)
			return
		}

		//1.5 Проверяем, есть ли активный реактор
		if user.ActiveReactor == 0 {
			http.Error(w, "no active reactor set", http.StatusBadRequest)
		}

		//2. Находим реактор
		var active_reactor models.Reactor
		err = db.Get(&active_reactor, "SELECT * FROM reactors WHERE id=$1", user.ActiveReactor)
		if err != nil {
			http.Error(w, "reactor not found", http.StatusInternalServerError)
			return
		}

		//3. Формируем ответ

		resp := BonusResponse{
			Balance:       user.Balance,
			ActiveReactor: active_reactor,
			FarmStatus:    user.FarmStatus,
			Progress:      user.FarmProgress,
			TimeLeft:      0,
		}

		//4. Обработка состояния фарминга
		if user.FarmStatus == "farming" && user.FarmStartTime != nil {
			elapsed := int(time.Since(*user.FarmStartTime).Seconds())
			total := active_reactor.FarmTime
			resp.Progress = min((elapsed*100)/total, 100)
			if elapsed >= active_reactor.FarmTime {
				resp.FarmStatus = "claim"
				resp.TimeLeft = 0
			} else {
				resp.TimeLeft = active_reactor.FarmTime - elapsed
			}
		}

		//5. отправка JSON
		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(resp)
	}
}

func StartFarming(db *sqlx.DB) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		var user models.User

		userIDStr := r.URL.Query().Get("userID")
		userID, err := strconv.Atoi(userIDStr)
		if err != nil {
			http.Error(w, "Invalid id", http.StatusBadRequest)
			return
		}

		//Находим пользователя
		err = db.Get(&user, "SELECT * FROM users WHERE id=$1", userID)
		if err != nil {
			http.Error(w, "User not found", http.StatusNotFound)
			return
		}

		//Проверяем, не фармит ли он уже
		if user.FarmStatus != "start" {
			http.Error(w, "User is already farming", http.StatusBadRequest)
			return
		}

		//Обновляем статус и время начала фарма

		_, err = db.Exec(`
		UPDATE users
		SET farm_status=$1, farm_start_time=$2
		WHERE id = $3`, "farming", time.Now(), userID)
		if err != nil {
			http.Error(w, "Failed to start farming", http.StatusInternalServerError)
			return
		}

		w.WriteHeader(http.StatusOK)
		w.Write([]byte("Farming started"))

	}
}

func ClaimBonuses(db *sqlx.DB) http.HandlerFunc { // /bonuses/claim
	return func(w http.ResponseWriter, r *http.Request) {
		var user models.User

		userIDStr := r.URL.Query().Get("userID")
		userID, err := strconv.Atoi(userIDStr)
		if err != nil {
			http.Error(w, "Invalid id", http.StatusBadRequest)
			return
		}

		err = db.Get(&user, "SELECT * FROM users WHERE id = $1", userID)
		if err != nil {
			http.Error(w, "User not found", http.StatusNotFound)
			return
		}

		//Находим активный реактор

		var reactor models.Reactor

		err = db.Get(&reactor, "SELECT * FROM reactors WHERE id = $1", user.ActiveReactor)
		if err != nil {
			http.Error(w, fmt.Sprintf("Reactor not found: %v", err), http.StatusNotFound)
			return
		}

		//Проверяем, может ли он заклеймить поинты

		if user.FarmStatus != "farming" || user.FarmStartTime.IsZero() {
			http.Error(w, "Farming not started of already claimed", http.StatusBadRequest)
			return
		}

		elapsed := int(time.Since(*user.FarmStartTime).Seconds())
		if elapsed < reactor.FarmTime {
			http.Error(w, "Farming not yet complete", http.StatusBadRequest)
			return
		}

		//Начисляем токены и сбрасываем статус
		newBalance := user.Balance + reactor.TokensPerCycle
		_, err = db.Exec(`
		UPDATE users
		SET balance=$1, farm_status = $2, farm_start_time=NULL
		WHERE id = $3`, newBalance, "start", userID)

		if err != nil {
			http.Error(w, "Failed to claim bonuses", http.StatusInternalServerError)
			return
		}

		w.WriteHeader(http.StatusOK)
		w.Write([]byte("Bonuses claimed"))
	}
}

File: reactor-game/frontend/src/stores/bonuses.ts
import { defineStore } from 'pinia'
import axios from 'axios'
import { reactive } from 'vue'
import {API_BASE_URL} from "../config"
// Интерфейс для реактора
interface Reactor {
  id: number
  farm_time: number
  tokens_per_cycle: number
  price: number
}

// Интерфейс для состояния
interface BonusState {
  balance: number
  active_reactor: Reactor
  farm_status: 'start' | 'farming' | 'claim'
  progress: number
  time_left: number
}

// Создаём стор
export const useBonusesStore = defineStore('bonuses', () => {
  // Реактивное состояние
  const state = reactive<BonusState>({
    balance: 0,
    active_reactor: { id: 0, farm_time: 0, tokens_per_cycle: 0, price: 0 },
    farm_status: 'start',
    progress: 0,
    time_left: 0
  })

  // Локальная переменная для таймера
  let timer: number = 0

  // Действия
  const fetchBonuses = async () => {
    try {
      const response = await axios.get<BonusState>(`${API_BASE_URL}/bonuses?userID=${localStorage.getItem('userID')}`)
      Object.assign(state, response.data)
      console.log(state)
      if (state.farm_status === 'farming') {
        startTimer()
      }
    } catch (err) {
      console.error('Failed to fetch bonuses:', err)
    }
  }

  const startFarming = async () => {
    if (state.farm_status !== 'start') return
    try {
      await axios.post(`${API_BASE_URL}/bonuses/start?userID=${localStorage.getItem('userID')}`)
      await fetchBonuses()
    } catch (err) {
      console.error('Failed to start farming:', err)
    }
  }

  const claimBonuses = async () => {
    if (state.farm_status !== 'claim') return
    try {
      await axios.post(`${API_BASE_URL}/bonuses/claim?userID=${localStorage.getItem('userID')}`)
      await fetchBonuses()
    } catch (error) {
      console.error('Failed to claim bonuses:', error)
    }
  }

  const startTimer = () => {
    if (timer) clearInterval(timer) 
    timer = setInterval(() => {
      if (state.time_left > 0) {
        state.time_left -= 1
        state.progress = Math.min(
          ((state.active_reactor.farm_time - state.time_left) / state.active_reactor.farm_time) * 100,
          100
        ) 
      }
      if (state.time_left <= 0 && state.farm_status === 'farming') {
        state.farm_status = 'claim'
        state.progress = 100
        clearInterval(timer) 
      }
    }, 1000)
  }

  const clearTimer = () => {
    if (timer) clearInterval(timer) 
  }

  // Возвращаем состояние и методы
  return {
    state,
    fetchBonuses,
    startFarming,
    claimBonuses,
    startTimer,
    clearTimer
  }
})